pkg/package-server/storage/packagemanifest/watcher.go

```diff
@@ -0,0 +1,82 @@
+package packagemanifest
+
+import (
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/packagemanifest/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/provider"
+)
+
+type Watcher struct {
+	namespace       string
+	name            string
+	resourceVersion string
+	labelSelector   labels.Selector
+
+	source provider.PackageManifestProvider
+
+	stop   chan struct{}
+	result chan watch.Event
+}
+
+var _ watch.Interface = &Watcher{}
+
+func NewWatcher(namespace, name, resourceVersion string, labelSelector labels.Selector, source provider.PackageManifestProvider) *Watcher {
+	return &Watcher{
+		namespace:       namespace,
+		resourceVersion: resourceVersion,
+		labelSelector:   labelSelector,
+		source:          source,
+		stop:            make(chan struct{}),
+		result:          make(chan watch.Event),
+	}
+}
+
+// Run is a blocking method which starts the watch by subscribing to the source's channels.
+// Should run in a goroutine.
+func (w *Watcher) Run() {
+	add, modify, delete := w.source.Subscribe()
+	for {
+		select {
+		case manifest := <-add:
+			w.Add(manifest)
+		case manifest := <-modify:
+			w.Modify(manifest)
+		case manifest := <-delete:
+			w.Delete(manifest)
+		case <-w.stop:
+			return
+		}
+	}
+}
+
+func (w *Watcher) Stop() {
```

Comment:
I'm pretty sure it's internal k8s code, because it's [part of the interface of `watch.Interface`](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/vendor/k8s.io/apimachinery/pkg/watch/watch.go#L290).

Reply:
> Do you know what calls `Stop()` on the watcher?
> 
> Also, probably want to guard against calling multiple times (double-close will panic):
> 
> ```go
> w.Lock()
> defer w.Unlock()
> if !w.Stopped {
>   close(w.result)
>   w.Stopped = true
> }
> ```

alternatively:

`w.once.Do(func() { close(w.result) })` where `w.once` is of type `sync.Once`

Reply:
@alecmerdler You're right, and the code that calls it is here: https://github.com/kubernetes/sample-apiserver/blob/df44042/vendor/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

As long as we're going through this codepath for the aggregated apiserver we're fine. Also note that it servers http or websockets :)